cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -287,26 +257,126 @@ public ArtifactInfo getArtifact(Id.Artifact artifactId) throws ArtifactNotExists
+      new TransactionExecutor.Function<DatasetContext<Table>, Map.Entry<ArtifactInfo, PluginClass>>() {
+        @Override
+        public Map.Entry<ArtifactInfo, PluginClass> apply(DatasetContext<Table> context) throws Exception {
+          PluginKey pluginKey = new PluginKey(artifactId.getNamespace(), type, name);
+          ArtifactColumn column = new ArtifactColumn(artifactId.getVersion().getVersion(), artifactId.getName());
+          byte[] value = context.get().get(pluginKey.getRowKey(), column.getColumn());
+          if (value == null) {
+            return null;
+          }
+          PluginData pluginData = gson.fromJson(Bytes.toString(value), PluginData.class);
+          return Maps.immutableEntry(new ArtifactInfo(artifactId, pluginData.artifactLocation), pluginData.pluginClass);
+        }
+      });
+    if (plugin == null) {
+      throw new PluginNotExistsException(artifactId, type, name);
+    }
+    return plugin;
+  }
+
+  /**
+   * Write the artifact and its metadata to the store. Once added, artifacts cannot be changed unless they are
+   * snapshot versions.
+   *
+   * @param artifactId the id of the artifact to add
+   * @param artifactMeta the metadata for the artifact
+   * @param artifactContents the contents of the artifact
+   * @throws WriteConflictException if the artifact is already currently being written
+   * @throws ArtifactAlreadyExistsException if a non-snapshot version of the artifact already exists
+   * @throws IOException if there was an exception persisting the artifact contents to the filesystem,
+   *                     of persisting the artifact metadata to the metastore
+   */
+  public void write(final Id.Artifact artifactId, final ArtifactMeta artifactMeta, final InputStream artifactContents)
+    throws WriteConflictException, ArtifactAlreadyExistsException, IOException {
+
+    Location fileDirectory =
+      locationFactory.get(artifactId.getNamespace(), ARTIFACTS_PATH).append(artifactId.getName());
+    Locations.mkdirsIfNotExists(fileDirectory);
+
+    // write the file contents
+    final Location destination = fileDirectory.append(artifactId.getVersion().getVersion()).getTempFile(".jar");
+    ByteStreams.copy(artifactContents, destination.getOutputStream());
+
+    // now try and write the metadata for the artifact
+    try {
+      boolean written = metaTable.execute(new TransactionExecutor.Function<DatasetContext<Table>, Boolean>() {
+
+        @Override
+        public Boolean apply(DatasetContext<Table> context) throws Exception {
+          Table table = context.get();
+
+          ArtifactCell artifactCell = new ArtifactCell(artifactId);
+          byte[] existingMetaBytes = table.get(artifactCell.rowkey, artifactCell.column);
+          boolean isSnapshot = artifactId.getVersion().isSnapshot();
+          if (existingMetaBytes != null && !isSnapshot) {
+            // non-snapshot artifacts are immutable. If there is existing metadata, stop here.
```

Comment:
So we will always copy the JAR to location irrespective of whether we finally choose to update the metadata or not?